### PR TITLE
Add OpenSearch metrics

### PR DESCRIPTION
**Pull request by noCharger**
## Description

This PR is to 
* add a rest high level client wrapper for operations around the original rest high level client using the Decorator Pattern, considering [RestHighLevelClient](https://github.com/opensearch-project/OpenSearch/blob/main/client/rest-high-level/src/main/java/org/opensearch/client/RestHighLevelClient.java#L262) itself is not extensible.
* add spark metrics source for flint
* integrate the client wrapper to be the aspect for collecting opensearch metrics

Metrics included in this PR:

```
OpenSearch Status Count, Dimension by default [applicationId, jobId, instanceRole]

opensearch.read.2xx.count
opensearch.read.4xx.count
opensearch.read.5xx.count
opensearch.read.403.count

opensearch.write.2xx.count
opensearch.write.4xx.count
opensearch.write.5xx.count
opensearch.write.403.count
```
Screenshot 2024-01-31 at 10 14 09 AM

## Approaches

There are three approaches to achieving aspect-oriented programming (AOP) to provide metrics within Flint: AspectJ library, dynamic proxies, and wrapper (decorator).

### AspectJ

- **AOP Implementation:** AspectJ is a robust AOP solution that enables the seamless and enhanced integration of cross-cutting concerns without altering the existing code base. It employs bytecode weaving (either at compile-time or load-time) to provide a wide range of AOP capabilities such as method interception, behavior modification, and even class structure change. PoC: https://github.com/opensearch-project/opensearch-spark/pull/219. 
- **Performance:** Because AspectJ alters the bytecode directly, method calls perform virtually natively. This makes it ideal for high-performance and high-throughput applications.
- **Use Cases**: Ideally suited for complicated applications with several cross-cutting concerns. AspectJ is especially handy when requiring advanced AOP features that extend beyond simple method interception.
- **Challenge**: The official sbt-aspectj plugin, https://github.com/sbt/sbt-aspectj, is less maintained. More effort is needed to adapt the AspectJ configuration for the Flint repository.

### Java Dynamic Proxies
**AOP Implementation**: Dynamic proxies make it easier to implement AOP by intercepting methods via reflection. They are limited to interfaces and cannot directly change class behavior or attributes.

Code example:

```
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.logging.Logger;

public class RestClientInvocationHandler implements InvocationHandler {
    private static final Logger LOG = Logger.getLogger(RestClientInvocationHandler.class.getName());
    private final Object target;

    public RestClientInvocationHandler(Object target) {
        this.target = target;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        long startTime = System.currentTimeMillis();
        try {
            Object result = method.invoke(target, args);
            logMetric("success", System.currentTimeMillis() - startTime);
            return result;
        } catch (Throwable t) {
            logMetric("failure", System.currentTimeMillis() - startTime);
            throw t;
        }
    }

    private void logMetric(String status, long duration) {
        LOG.info("Method " + status + ": " + duration + "ms");
    }
}

import org.opensearch.client.RestHighLevelClient;
import java.lang.reflect.Proxy;

public class RestClientProxyFactory {

    public static RestHighLevelClient createProxy(RestHighLevelClient client) {
        return (RestHighLevelClient) Proxy.newProxyInstance(
            client.getClass().getClassLoader(),
            new Class<?>[]{RestHighLevelClient.class},
            new RestClientInvocationHandler(client)
        );
    }
}


RestHighLevelClient realClient = new RestHighLevelClient(...);
RestHighLevelClient client = RestClientProxyFactory.createProxy(realClient);
```

**Performance**: Because of the usage of reflection, they have a larger overhead than AspectJ approach. However, for many common applications, this expense is insignificant.
**Use cases**: Suitable for applications in which the AOP requirements are confined to method interception and interfaces are the norm. It's a suitable alternative for basic AOP scenarios and projects where introducing a dependency like AspectJ isn't ideal.

### Wrapper (Decorator) Approach implemented in this PR
**AOP Implementation:** This approach doesn't use AOP in the traditional sense but can mimic some aspects of it, like method interception, by manually wrapping classes. It's more of a design pattern than an AOP solution and requires explicit coding for each method and class.
**Performance:** The wrapper approach introduces additional method calls, but these are usually direct and thus have less overhead than dynamic proxies. The performance impact is generally minimal, but it's more than AspectJ.
**Use Cases:** Best for situations where you need to wrap a limited number of classes or methods, and the interface of the classes is stable. It's straightforward and easy to understand, making it a good choice for smaller projects or when simplicity is key.

Overall Comparison
**Performance Hierarchy:** AspectJ > Wrapper Approach > Dynamic Proxies. AspectJ is typically the fastest due to direct bytecode manipulation, followed by the wrapper approach with its direct method calls, and finally dynamic proxies with the overhead of reflection.
**Complexity and Flexibility:** AspectJ is the most complex but also the most flexible and powerful. Dynamic proxies offer a balance between simplicity and flexibility, suitable for medium complexity needs. The wrapper approach is the simplest but the least flexible, especially for large-scale or changing codebases.
**Suitability:** AspectJ is ideal for large, complex applications with extensive cross-cutting concerns. Dynamic proxies are good for medium-scale applications or where interface-based programming is predominant. The wrapper approach is best for smaller applications or when you need a straightforward solution without extra dependencies or complexities.

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/220

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-core/src/main/scala/org/opensearch/flint/core/metrics/IRestHighLevelClient.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package org.opensearch.flint.core.metrics;
+
+public interface IRestHighLevelClient {
```
Comment (penghuo) [inline flint-core/src/main/scala/org/opensearch/flint/core/metrics/IRestHighLevelClient.java:3]: Why IRestHighLevelClient is required?

File: flint-core/src/main/scala/org/opensearch/flint/core/metrics/RestHighLevelClientWrapper.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import org.opensearch.client.indices.GetIndexRequest;
+import org.opensearch.client.indices.GetIndexResponse;
+import org.opensearch.action.admin.indices.delete.DeleteIndexRequest;
+import org.opensearch.client.RequestOptions;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+public class RestHighLevelClientWrapper implements IRestHighLevelClient, Closeable {
```
Comment (penghuo) [inline flint-core/src/main/scala/org/opensearch/flint/core/metrics/RestHighLevelClientWrapper.java:30]: Is it possible to publish metrics in RetryableHttpAsyncClient? @dai-chen  any suggestion?

Comment (dai-chen) [reply flint-core/src/main/scala/org/opensearch/flint/core/metrics/RestHighLevelClientWrapper.java:30]: I think it's possible but there is only low-level HTTP request and response.
It depends on what metrics we want to publish, ex. FlintClient API metrics, OS client API metrics or HTTP request metrics etc

Comment (noCharger) [reply flint-core/src/main/scala/org/opensearch/flint/core/metrics/RestHighLevelClientWrapper.java:30]: > I think it's possible but there is only low-level HTTP request and response. It depends on what metrics we want to publish, ex. FlintClient API metrics, OS client API metrics or HTTP request metrics etc

+1, `RetryableHttpAsyncClient` is suitable if we have use case around HTTP req/res like `timeout`. @penghuo what's your suggestion?

Comment (noCharger) [reply flint-core/src/main/scala/org/opensearch/flint/core/metrics/RestHighLevelClientWrapper.java]: Discussed with @penghuo offline and we agreed to use the client wrapper approach

File: flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java [hunk 3]
```diff
@@ -96,8 +95,8 @@ <T> OptimisticTransaction<T> startTransaction(String indexName, String dataSourc
    */
   FlintWriter createWriter(String indexName);
   /**
-   * Create {@link RestHighLevelClient}.
-   * @return {@link RestHighLevelClient}
+   * Create {@link RestHighLevelClientWrapper}.
+   * @return {@link RestHighLevelClientWrapper}
    */
-  public RestHighLevelClient createClient();
+  public RestHighLevelClientWrapper createClient();
```
Comment (dai-chen) [inline flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java:101]: Should we use interface `IRestHighLevelClient` here?

File: spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala [hunk 4]
```diff
@@ -49,7 +49,7 @@ object FlintREPL extends Logging with FlintJobExecutor {
   val INITIAL_DELAY_MILLIS = 3000L
   val EARLY_TERMIANTION_CHECK_FREQUENCY = 60000L
 
-  def update(flintCommand: FlintCommand, updater: OpenSearchUpdater): Unit = {
+  def updateSessionIndex(flintCommand: FlintCommand, updater: OpenSearchUpdater): Unit = {
```
[the remark this answers is not in the document]
Comment (noCharger) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:52]: Correct, unrelated to metrics. I would like to change the method name to be more specific, so we can identify lower level meta storage operations from high level repl actions.

[the remark this answers is not in the document]
Comment (noCharger) [reply spark-sql-application/src/main/scala/org/apache/spark/sql/FlintREPL.scala:52]: > Correct, unrelated to metrics. I would like to change the method name to be more specific, so we can identify lower level meta storage operations from high level repl actions.

Correct, unrelated to metrics. I would like to change the method name to be more specific, so we can identify lower level meta storage operations from high level repl actions.

File: flint-core/src/main/java/org/opensearch/flint/core/IRestHighLevelClient.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import org.opensearch.action.admin.indices.delete.DeleteIndexRequest;
+import org.opensearch.client.RequestOptions;
+
+import java.io.IOException;
+
+/**
+ * Interface for wrapping the OpenSearch High Level REST Client with additional functionality,
+ * such as metrics tracking.
+ */
+public interface IRestHighLevelClient {
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/IRestHighLevelClient.java]: Is this consolidated list or the only methods that we are using?

File: flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java (hunk 3 above)
Comment (noCharger) [reply flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java]: Revised in https://github.com/opensearch-project/opensearch-spark/pull/229/commits/ece16a0bfc2ee466b815b98a3d4629a7d23ca665

File: flint-core/src/main/java/org/opensearch/flint/core/IRestHighLevelClient.java (hunk 5 above)
Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/IRestHighLevelClient.java]: Only the only methods that we are using.

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    // Constructs the metric name based on the provided prefix and status code
+    private static String constructMetricName(String metricNamePrefix, int statusCode) {
+        String metricSuffix = getMetricSuffixForStatusCode(statusCode);
+        return metricNamePrefix + "." + metricSuffix;
+    }
+
+    // Determines the metric suffix based on the HTTP status code
+    private static String getMetricSuffixForStatusCode(int statusCode) {
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:49]: Can we add in the description what all metrics are we covering with this PR.

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        if (statusCode == 403) {
+            return "403.count";
+        } else if (statusCode >= 500) {
+            return "5xx.count";
+        } else if (statusCode >= 400) {
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:54]: 4xx is all but 403?

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+     *
+     * @param client the RestHighLevelClient instance to wrap
+     */
+    public RestHighLevelClientWrapper(RestHighLevelClient client) {
+        this.client = client;
+    }
+
+    @Override
+    public BulkResponse bulk(BulkRequest bulkRequest, RequestOptions options) throws IOException {
+        return execute(OS_WRITE_OP_METRIC_PREFIX, () -> client.bulk(bulkRequest, options));
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:56]: Don't we want to add any dimensions along with it?
Also we have DimensionedName class which we can make use to add dimensions during metric publishing time.

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+     * Extracts an OpenSearchException from the given Throwable.
+     * Checks if the Throwable is an instance of OpenSearchException or caused by one.
+     *
+     * @param ex the exception to be checked
+     * @return the extracted OpenSearchException, or null if not found
+     */
+    private OpenSearchException extractOpenSearchException(Throwable ex) {
+        if (ex instanceof OpenSearchException) {
+            return (OpenSearchException) ex;
+        } else if (ex.getCause() instanceof OpenSearchException) {
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:145]: Whats the rationale behind this block?

Comment (vamsi-amazon) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:145]: In what scenarios OpenSearchException is encapsulated under OpenSearchException? Did you arrive at this after some testing?

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java (hunk 7 above)
Comment (vamsi-amazon) [reply flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:54]: can we reduce the logic to (statusCode/100)xx.count

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/reporter/DimensionedCloudWatchReporter.java [hunk 10]
```diff
@@ -366,36 +365,39 @@ private void stageMetricDatum(final boolean metricConfigured,
...
-        dimensions.add(instanceRoleDimension);
-        dimensions.add(domainIdDimension);
+        addDefaultDimensionsForSparkJobMetrics(dimensions);
         dimensions.addAll(dimensionedName.getDimensions());
         return new MetricInfo(metricName, dimensions);
     }
 
+    // These dimensions are for all metrics
+    private static void addDefaultDimensionsForSparkJobMetrics(Set<Dimension> dimensions) {
+        final String jobId = System.getenv().getOrDefault("SERVERLESS_EMR_JOB_ID", UNKNOWN);
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/reporter/DimensionedCloudWatchReporter.java:386]: As discussed offline, we can try removing as the library is not specific to emr. But that can be handled in another PR.

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/reporter/DimensionedCloudWatchReporter.java [hunk 11]
```diff
@@ -366,36 +365,39 @@ private void stageMetricDatum(final boolean metricConfigured,
...
-            //For executors only id is added to the metric name, that's why the numeric check.
-            //If it is not numeric then the instance is driver.
-            if (StringUtils.isNumeric(parts[1])) {
-                instanceRoleDimension = new Dimension().withName(DIMENSION_INSTANCE_ROLE).withValue("executor" + parts[1]);
-            }
-            else {
-                instanceRoleDimension = new Dimension().withName(DIMENSION_INSTANCE_ROLE).withValue(parts[1]);
-            }
+            // Trim custom metric source name
+            int skip = (parts[2].equals("Flint")) ? 3 : 2;
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/reporter/DimensionedCloudWatchReporter.java:375]: What is this logic?
Can we move this to a function with descriptive name?
Also thanks for refactoring.

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    private MetricsUtil() {
+    }
+
+    /**
+     * Publish an OpenSearch metric based on the status code.
+     *
+     * @param metricNamePrefix the prefix for the metric name
+     * @param statusCode       the HTTP status code
+     */
+    public static void publishOpenSearchMetric(String metricNamePrefix, int statusCode) {
```
Comment (vamsi-amazon) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:34]: publishOpenSearchStatusMetric??

**Comment by vamsi-amazon** [comment]
Thanks for the changes.

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java (hunk 9 above)
Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:177]: > In what scenarios OpenSearchException is encapsulated under OpenSearchException? Did you arrive at this after some testing?

https://github.com/opensearch-project/OpenSearch/blob/92afd67c570a779cb33846c9359b0cbd6fe56753/client/rest/src/main/java/org/opensearch/client/RestClient.java#L342 If I understand correctly, the rest client can wrap OpenSearchException.

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+ */
+public final class MetricsUtil {
+
+    private static final Logger LOG = Logger.getLogger(MetricsUtil.class.getName());
+
+    // Private constructor to prevent instantiation
+    private MetricsUtil() {
+    }
+
+    public static void incrementCounter(String metricName) {
```
Comment (penghuo) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:27]: add doc to explain the function of this public interface

File: flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        return counter;
+    }
+
+    // Gets or initializes the FlintMetricSource
+    private static FlintMetricSource getOrInitFlintMetricSource(SparkEnv sparkEnv) {
+        Seq<Source> metricSourceSeq = sparkEnv.metricsSystem().getSourcesByName(FlintMetricSource.FLINT_METRIC_SOURCE_NAME());
+
+        if (metricSourceSeq == null || metricSourceSeq.isEmpty()) {
+            FlintMetricSource metricSource = new FlintMetricSource();
+            sparkEnv.metricsSystem().registerSource(metricSource);
```
Comment (penghuo) [inline flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:56]: FlintMetricSource is registered only first time incrementCounter() is called, is it possible metrics should be collected without calling incrementCount(), for instance, driverCount / executorCount

Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/metrics/MetricsUtil.java:56]: > FlintMetricSource is registered only first time incrementCounter() is called

The Spark MetricsSystem will register all sinks/sources during `start()` lifecycle and register all sources defined in config. The FlintMetricSource is created via reflection.

https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala#L96-L105
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala#L179-L193

> is it possible metrics should be collected without calling incrementCount(), for instance, driverCount / executorCount

The Spark internal source like ExecutetAllocationManager (for executorCount) is created during SparkContext Initialization.

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+     * Otherwise, it increments a general failure metric counter based on the status code category (e.g., 4xx, 5xx).
+     *
+     * @param metricNamePrefix the prefix for the metric name which is used to construct the full metric name for failure
+     * @param e                the exception encountered during the operation, used to determine the type of failure
+     */
+    private void recordOperationFailure(String metricNamePrefix, Exception e) {
+        OpenSearchException openSearchException = extractOpenSearchException(e);
+        int statusCode = openSearchException != null ? openSearchException.status().getStatus() : 500;
+
+        if (statusCode == 403) {
```
[the remark this answers is not in the document]
Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:158]: Considering [this](https://github.com/opensearch-project/opensearch-spark/pull/229#discussion_r1470037799), we will have [2xx, 403, 4xx, 5xx]

[the remark this answers is not in the document]
Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:158]: Also updated in decription.

**Review by penghuo — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Review by vamsi-amazon — APPROVED** [review]

File: flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java (hunk 9 above)
Comment (noCharger) [reply flint-core/src/main/java/org/opensearch/flint/core/RestHighLevelClientWrapper.java:177]: Reproduced when the user doesn't any opensearch permission:

```
Exception in thread "main" OpenSearchStatusException[method [HEAD], host [***], URI [/flint_ql_sessions?ignore_throttled=false&ignore_unavailable=false&expand_wildcards=open%2Cclosed&allow_no_indices=false], status line [HTTP/1.1 403 Forbidden]]; nested: ResponseException[method [HEAD], host [***], URI [/flint_ql_sessions?ignore_throttled=false&ignore_unavailable=false&expand_wildcards=open%2Cclosed&allow_no_indices=false], status line [HTTP/1.1 403 Forbidden]];
	at org.opensearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:2202)
	at org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1924)
	at org.opensearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1894)
	at org.opensearch.client.IndicesClient.exists(IndicesClient.java:920)
	at org.opensearch.flint.core.RestHighLevelClientWrapper.lambda$isIndexExists$8(RestHighLevelClientWrapper.java:95)
	at org.opensearch.flint.core.RestHighLevelClientWrapper.execute(RestHighLevelClientWrapper.java:124)
	at org.opensearch.flint.core.RestHighLevelClientWrapper.isIndexExists(RestHighLevelClientWrapper.java:95)
	at org.opensearch.flint.core.storage.OpenSearchUpdater.assertIndexExist(OpenSearchUpdater.java:85)
	at org.opensearch.flint.core.storage.OpenSearchUpdater.upsert(OpenSearchUpdater.java:34)
	at org.apache.spark.sql.FlintREPL$.updateFlintInstance(FlintREPL.scala:431)
	at org.apache.spark.sql.FlintREPL$.handleSessionError(FlintREPL.scala:390)
	at org.apache.spark.sql.FlintREPL$.main(FlintREPL.scala:171)
	at org.apache.spark.sql.FlintREPL.main(FlintREPL.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1006)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1095)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1104)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: org.opensearch.client.ResponseException: method [HEAD], host [***], URI [/flint_ql_sessions?ignore_throttled=false&ignore_unavailable=false&expand_wildcards=open%2Cclosed&allow_no_indices=false], status line [HTTP/1.1 403 Forbidden]
	at org.opensearch.client.RestClient.convertResponse(RestClient.java:375)
	at org.opensearch.client.RestClient.performRequest(RestClient.java:345)
	at org.opensearch.client.RestClient.performRequest(RestClient.java:320)
	at org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1911)
	... 23 more
```

Screenshot 2024-02-14 at 3 26 51 PM

**Pull request merged**